clang/lib/CodeGen/CodeGenAction.cpp

```diff
@@ -825,6 +826,14 @@ void BackendConsumer::DontCallDiagHandler(const DiagnosticInfoDontCall &D) {
       << llvm::demangle(D.getFunctionName().str()) << D.getNote();
 }
 
+void BackendConsumer::SYCLWarningDiagHandler(
+    const DiagnosticInfoSYCLUnspecAspect &D) {
+  SourceLocation Loc;
+  Diags.Report(Loc, diag::warn_sycl_wrong_aspect_usage)
```

Comment:
Oh, it uses invalid source location, so the warning won't point to real code. Maybe we can try to construct valid source info at least when we have debug info? Seems like `DontCallDiagHandler`  defined above at least tries something like that.

Reply:
This is a good idea and it seems like that is a better approach than we have right now. However, `DontCallDiagHandler` relies on source location emitted into LLVM IR by FE in form of `srcloc` metadata: [D106030](https://reviews.llvm.org/D106030), [D110364](https://reviews.llvm.org/D110364).

Amount of the changes in this PR is already huge, would it be acceptable to postpone this diagnostic improvement till the next PR?

Reply:
> Amount of the changes in this PR is already huge, would it be acceptable to postpone this diagnostic improvement till the next PR?

I think it is ok to postpone. Please file a tracker if you're not going to work on it ASAP.

Reply:
> I'm less certain it's okay to postpone as this means all diagnostics will show up in very wrong places (if at all) within an IDE, for example. I don't think the user experience is all that great when the diagnostic is a warning that's divorced from their code.

Note that we are in the very beginning of our optional kernel features implementation journey and even if user properly sets all the attributes, the code won't work as expected, because we are missing a lot of changes in other components. Therefore, I think that it should be ok not to complicate this PR even more, but still refactor diagnostics _before_ we mark the whole feature as completed.

> Note: losing source location information and providing a poor user experience with diagnostics is why very few diagnostics ever come from the backend. It's too easy to lose the source locations, especially when optimizer passes get involved. That's why the majority of backend diagnostics are in the form of remarks instead of warnings (let alone errors). We should be very careful about the user experience here.

That pass is supposed to be launched before all other passes or otherwise not only source locations will be lost, but also some semantics, i.e. we can lose information about declared aspects due to inlining.

Reply:
> Therefore, I think that it should be ok not to complicate this PR even more, but still refactor diagnostics before we mark the whole feature as completed.

I'm ok with this approach.

Reply:
I've created #5877 to track this further improvement